narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,163 @@
 // Copyright (c) 2021, Facebook, Inc. and its affiliates
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use config::AuthorityIdentifier;
+use config::{AuthorityIdentifier, Committee};
+use crypto::RandomnessPrivateKey;
+use fastcrypto::groups;
+use fastcrypto_tbls::dkg;
 use mysten_metrics::metered_channel::{Receiver, Sender};
 use mysten_metrics::spawn_logged_monitored_task;
 use tap::TapFallible;
 use tokio::task::JoinHandle;
 use tracing::{debug, error, info, warn};
-use types::{Certificate, CertificateAPI, ConditionalBroadcastReceiver, HeaderAPI, Round};
+use types::{
+    Certificate, CertificateAPI, ConditionalBroadcastReceiver, HeaderAPI, Round, SystemMessage,
+};
 
-/// Receives the highest round reached by consensus and update it for all tasks.
+type G = groups::bls12381::G2Element; // TODO-DNS is it correct for these to be the same?
+type EG = groups::bls12381::G2Element;
+
+/// Updates Narwhal system state based on certificates received from consensus.
 pub struct StateHandler {
-    /// The id of this authority.
     authority_id: AuthorityIdentifier,
+
     /// Receives the ordered certificates from consensus.
     rx_committed_certificates: Receiver<(Round, Vec<Certificate>)>,
     /// Channel to signal committee changes.
     rx_shutdown: ConditionalBroadcastReceiver,
     /// A channel to update the committed rounds
     tx_committed_own_headers: Option<Sender<(Round, Vec<Round>)>>,
+    /// A channel to send system messages to the proposer.
+    tx_system_messages: Sender<SystemMessage>,
+
+    /// If set, generates Narwhal system messages for random beacn
+    /// DKG and randomness generation.
+    randomness_state: Option<RandomnessState>,
 
     network: anemo::Network,
 }
 
+// Internal state for randomness DKG and generation.
+struct RandomnessState {
+    party: dkg::Party<G, EG>,
+    messages: Vec<dkg::Message<G, EG>>,
+    processed_messages: Vec<dkg::ProcessedMessage<G, EG>>,
+    shares: dkg::SharesMap<<G as groups::GroupElement>::ScalarType>,
+    confirmations: Vec<dkg::Confirmation<EG>>,
+    dkg_output: Option<dkg::Output<G, EG>>,
+}
+
+impl RandomnessState {
+    fn try_new(committee: Committee, private_key: RandomnessPrivateKey) -> Option<Self> {
+        let total_stake = committee.total_stake() as f64;
+        let info = committee.randomness_dkg_info();
+
+        let nodes = info
+            .iter()
+            .map(|(id, pk, stake)| fastcrypto_tbls::nodes::Node::<EG> {
+                id: id.0,
```

Comment:
are those consecutive ids 1..n? the current interface requires that

Reply:
see https://github.com/MystenLabs/fastcrypto/blob/main/fastcrypto-tbls/src/dkg.rs#L36-L37
we also need to maintain something like "NW id -> DKG id" map, so we could verify that a DKG message is coming from the NW authenticated sender.

Reply:
It looks like it's 0..n-1 and not 1..n. Why does the protocol require 1..n? For that matter, why does it require consective IDs at all? Shouldn't any opaque identifier work, as long as it's consistent across validators?

If this really must be consecutive IDs 1..n it would be ideal if you could enforce that via the type system on fastcrypto side. (For example by removing `id` as an input field in `Node` and instead handling ID assignment internally based on the order of the `Vec` passed to `Nodes::new`.)